# This is a shell script to get intergenic regions from a list of intergenic mutations generated by snippy, annoted them from the gff of the reference and get the  # fasta sequences
# It takes one positional argument: the directory with the masked snippy output (output of mask-variants.sh)
# Note that it needs a conda snippy environment

# Functions
usage () {
	echo "USAGE"
        echo "  $0 [options] <grouped snipp-mask directory>"
        echo "OPTIONS"
        echo "  -f (force): remove existing directory"
	echo "  -i (input-file): file with clean mutations"
}

# Run script
if [ $# -eq 0 ]
then
	usage
	exit 1
fi

# Parse options
FORCE=false

while getopts 'fi:' option
do
  	case $option in
                f) FORCE=true ;;
		i) MUTATIONS_FILE=$(readlink -e $OPTARG) ;;
        esac
done

# skip to parse command line arguments
shift $((OPTIND-1))

# define variables
dir=$(readlink -e $1)

# create new directory
new_dir=$(basename $dir)

if [ -d $new_dir ]
then 
	if $FORCE
	then
		echo "Removing existing $new_dir directory"
		rm -r $new_dir
	else
		echo "Directory $new_dir exists. Use -f to remove directory. Exiting"
               	exit 1
	fi
fi

mkdir $new_dir
cd $new_dir

# Get intergenic regions of the reference

mkdir reference

cd reference

cp $dir/ref.fa .
cp $dir/ref.gff .
cp $dir/ref.gbk .
#genbank2fasta.pl --prot ref.gbk > ref.faa
grep CDS ref.gff > ref.cds.gff

samtools faidx ref.fa
bedtools makewindows -g ref.fa.fai -n 1 | bedtools subtract -a stdin -b ref.cds.gff > ref.intergenic.bed

cd ..

# Get interegenic regions of intergenic mutations

if [ -v MUTATIONS_FILE ]
then 
	echo "Extracting mutations from clean mutation file $MUTATIONS_FILE"
	grep $new_dir $MUTATIONS_FILE  | grep -v CDS | grep -v CHROM | cut -f 4,5 | sort -u | awk -F '\t' 'BEGIN {OFS=FS}{print $1, $2-1, $2}' > snps.intergenic.bed
else
	MUTATIONS_FILE="$dir/all_snps.mask.tab"
	echo "Extracting mutations from snippy-mask file $MUTATIONS_FILE"
	cp $MUTATIONS_FILE .
	grep -v CDS $MUTATIONS_FILE | cut -f4,5 | sed '1d' | sort -u | awk -F '\t' 'BEGIN {OFS=FS}{print $1, $2-1, $2}' > snps.intergenic.bed
fi

bedtools intersect -a reference/ref.intergenic.bed -b snps.intergenic.bed -wa -wb > mutated_intergenic_regions.bed
bedtools getfasta -fi reference/ref.fa -bed mutated_intergenic_regions.bed > mutated_intergenic_regions.fa
bedtools window -a reference/ref.cds.gff -b mutated_intergenic_regions.bed -w 1 | sed "s/^/$new_dir\t/" >  mutated_intergenic_regions_annotated.tab

# get flanking genes
#cut -f10 mutated_intergenic_regions_annotated.tab | cut -f1 -d ";" | cut -f2 -d "=" > flanking_proteins.txt
#for i in $(cat flanking_proteins.txt)
#	do fa-grep.pl --i $i reference/ref.faa
#done > flanking_proteins.faa

cd ..


